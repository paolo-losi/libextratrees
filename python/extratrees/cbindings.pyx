import numpy
cimport cython
cimport numpy as np
from libc.stdlib cimport malloc, free
from libc.stdint cimport uint32_t
from libc cimport math
from libcpp cimport bool
from cextratrees cimport (ET_problem, ET_problem_destroy,
                          ET_forest, ET_forest_destroy, ET_forest_build,
                          ET_forest_predict, ET_forest_predict_regression,
                          ET_forest_predict_class_majority,
                          ET_forest_predict_probability,
                          ET_forest_neighbors, ET_params,
                          ET_forest_predict_class_bayes,
                          class_probability_vec, class_probability,
                          neighbour_weight, neighbour_weight_vec,
                          double_vec, ET_forest_feature_importance)


cdef class Problem:

    cdef ET_problem *_prob

    def __dealloc__(self):
        if self._prob:
            ET_problem_destroy(self._prob)
            free(self._prob)

    def _train(self, number_of_features_tested=None,
                     number_of_trees=100,
                     regression=False,
                     min_split_size=1,
                     select_features_with_replacement=False):
        cdef ET_params params
        cdef ET_forest *cforest
        cdef ET_problem *cprob = self._prob

        if number_of_features_tested is None:
            if regression:
                number_of_features_tested = cprob.n_features
            else:
                number_of_features_tested = math.ceil(
                                                math.sqrt(cprob.n_features))

        params.number_of_features_tested = number_of_features_tested
        params.number_of_trees = number_of_trees
        params.regression = regression
        params.min_split_size = min_split_size
        params.select_features_with_replacement = \
                                             select_features_with_replacement

        cforest = ET_forest_build(cprob, &params)
        return forest_factory(cforest)


cdef Problem problem_factory(ET_problem *prob):
    cdef Problem instance = Problem.__new__(Problem)
    instance._prob = prob
    return instance


cdef double _p_simple(ET_forest *f, float *v, uint32_t curtail, bool _):
    return ET_forest_predict(f, v)

cdef double _p_regression(ET_forest *f, float *v, uint32_t curtail, bool _):
    return ET_forest_predict_regression(f, v, curtail)

cdef double _p_class_majority(ET_forest *f, float *v, uint32_t curtail, bool _):
    return ET_forest_predict_class_majority(f, v, curtail)

cdef double _p_cl_bayes(ET_forest *f, float *v, uint32_t curtail, bool smooth):
    return ET_forest_predict_class_bayes(f, v, curtail, smooth)


cdef class Forest:

    cdef ET_forest *_forest

    def __dealloc__(self):
        if self._forest:
            ET_forest_destroy(self._forest)
            free(self._forest)

    @cython.boundscheck(False)
    @cython.wraparound(False)
    def predict(self, np.ndarray[np.float32_t, ndim=2] X, bytes mode=None,
                curtail=1, smooth=False):
        cdef np.ndarray[np.float64_t, ndim=1] y
        cdef float *vector
        cdef int sample_idx, feature_idx
        cdef uint32_t _curtail = curtail
        cdef bool _smooth = smooth
        cdef double (*predict_f)(ET_forest *f, float *v, uint32_t c, bool s)

        if mode is None:
            predict_f = _p_simple
        elif mode == 'regression':
            predict_f = _p_regression
        elif mode == 'classify_majority':
            predict_f = _p_class_majority
        elif mode == 'classify_bayes':
            predict_f = _p_cl_bayes
        else:
            raise ValueError('unsupported predict mode: %r' % mode)

        if mode is None and curtail != 1:
            raise ValueError('curtail not supported for simple prediction')

        if mode != 'classify_bayes' and smooth == True:
            raise ValueError('smooth supported only for "classify_bayes" mode')

        y = numpy.empty(shape=(X.shape[0],), dtype=numpy.float64)
        vector = <float *> malloc(sizeof(float) * X.shape[1])
        if not vector:
            raise MemoryError()

        for sample_idx in xrange(X.shape[0]):
            for feature_idx in xrange(X.shape[1]):
                vector[feature_idx] = X[sample_idx, feature_idx]

            y[sample_idx] = predict_f(self._forest, vector, _curtail, _smooth)

        free(vector)
        return y

    @cython.boundscheck(False)
    @cython.wraparound(False)
    def predict_proba(self, np.ndarray[np.float32_t, ndim=2] X,
                      curtail=1, smooth=False):
        cdef float *vector
        cdef int sample_idx, feature_idx
        cdef uint32_t _curtail = curtail
        cdef bool _smooth = smooth
        cdef class_probability_vec *cpv
        cdef class_probability *cp
        cdef np.ndarray[np.float64_t, ndim=1] classes = None
        cdef np.ndarray[np.float64_t, ndim=2] probas = None

        vector = <float *> malloc(sizeof(float) * X.shape[1])
        if not vector:
            raise MemoryError()

        for sample_idx in xrange(X.shape[0]):
            for feature_idx in xrange(X.shape[1]):
                vector[feature_idx] = X[sample_idx, feature_idx]

            cpv = ET_forest_predict_probability(self._forest, vector,
                                                _curtail, _smooth)
            if not cpv:
                raise MemoryError()

            if classes is None:
                shape = (X.shape[0], cpv.n)
                probas  = numpy.empty(shape=shape, dtype=numpy.float64)
                classes = numpy.empty(shape=(cpv.n,), dtype=numpy.float64)
                for i in xrange(cpv.n):
                    cp = &cpv.a[i]
                    classes[i] = cp.label
                    probas[sample_idx, i]  = cp.probability
            else:
                for i in xrange(cpv.n):
                    cp = &cpv.a[i]
                    probas[sample_idx, i]  = cp.probability

            free(cpv.a)
            free(cpv)

        free(vector)
        return classes, probas

    @cython.boundscheck(False)
    @cython.wraparound(False)
    def neighbors(self, np.ndarray[np.float32_t, ndim=2] X, curtail=1):
        cdef float *vector
        cdef int sample_idx, feature_idx
        cdef uint32_t _curtail = curtail
        cdef neighbour_weight_vec *nwv
        cdef neighbour_weight *nw
        cdef np.ndarray[np.float64_t, ndim=2] adiacency

        adiacency = numpy.zeros(shape=(X.shape[0], self._forest.n_samples),
                                dtype=numpy.float64)

        vector = <float *> malloc(sizeof(float) * X.shape[1])
        if not vector:
            raise MemoryError()

        for sample_idx in xrange(X.shape[0]):
            for feature_idx in xrange(X.shape[1]):
                vector[feature_idx] = X[sample_idx, feature_idx]

            nwv = ET_forest_neighbors(self._forest, vector, _curtail)
            if not nwv:
                raise MemoryError()

            for i in xrange(nwv.n):
                nw = &nwv.a[i]
                adiacency[sample_idx, nw.key] = nw.weight

            free(nwv.a)
            free(nwv)

        free(vector)
        return adiacency

    @cython.boundscheck(False)
    @cython.wraparound(False)
    def feature_importance(self):
        cdef uint32_t n_features = self._forest.n_features
        cdef double_vec *c_feat_imp = ET_forest_feature_importance(self._forest)
        cdef np.ndarray[np.float64_t, ndim=1] feat_imp

        if c_feat_imp == NULL:
            raise MemoryError()

        feat_imp = numpy.empty(shape=(n_features,))
        for i in xrange(n_features):
            feat_imp[i] = c_feat_imp.a[i]

        free(c_feat_imp.a)
        free(c_feat_imp)

        return feat_imp


cdef Forest forest_factory(ET_forest *forest):
    cdef Forest instance = Forest.__new__(Forest)
    instance._forest = forest
    return instance


# TODO use cython typed memoryviews to avoid copying
@cython.boundscheck(False)
@cython.wraparound(False)
def convert_to_problem(np.ndarray[np.float32_t, ndim=2] X not None,
                       np.ndarray[np.float64_t, ndim=1] y not None):

    if y.shape[0] != X.shape[0]:
        raise ValueError('y.shape[0] != X.shape[0]')

    cdef size_t n_features, n_samples
    cdef int i, j

    cdef ET_problem *cprob = <ET_problem *> malloc(sizeof(ET_problem))

    n_features = X.shape[1]
    n_samples  = X.shape[0]

    cprob.n_features = n_features
    cprob.n_samples = n_samples

    cprob.vectors = <float *> malloc(sizeof(float) * n_features * n_samples)
    cprob.labels = <double *> malloc(sizeof(double) * n_samples)
    if not cprob.labels or not cprob.vectors:
        raise MemoryError()

    for j in xrange(n_features):
        for i in xrange(n_samples):
            cprob.vectors[j * n_samples + i] = X[i, j]

    for i in xrange(n_samples):
        cprob.labels[i] = y[i]

    return problem_factory(cprob)


@cython.boundscheck(False)
@cython.wraparound(False)
def convert_from_problem(Problem prob not None):
    cdef int i, j
    cdef ET_problem *cprob = prob._prob
    cdef np.ndarray[np.float32_t, ndim=2] X
    cdef np.ndarray[np.float64_t, ndim=1] y

    X = numpy.empty(shape=(cprob.n_samples, cprob.n_features), dtype=np.float64)
    y = numpy.empty(shape=(cprob.n_samples,), dtype=np.float64)

    for j in xrange(cprob.n_features):
        for i in xrange(cprob.n_samples):
            X[i, j] = cprob.vectors[j * cprob.n_samples + i]

    for i in xrange(cprob.n_samples):
        y[i] = cprob.labels[i]

    return X, y


def train(X, y, **params):
    return convert_to_problem(X, y)._train(**params)


