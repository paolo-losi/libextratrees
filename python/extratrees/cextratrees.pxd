from libcpp cimport bool
from libc.stdint cimport uint32_t


cdef extern from "extratrees.h":
    
    ctypedef struct ET_problem:
        float *vectors
        double *labels
        uint32_t n_features
        uint32_t n_samples

    cdef void ET_problem_destroy(ET_problem *prob)
    cdef ET_problem *ET_load_libsvm_file(char *fname)

    ctypedef struct ET_base_node:
        pass
    
    ctypedef ET_base_node *ET_tree

    ctypedef struct tree_vec:
        size_t n, m
        ET_tree *a

    ctypedef struct ET_forest:
        uint32_t n_features
        uint32_t n_samples
        tree_vec trees

    cdef void ET_forest_destroy(ET_forest *forest)
    cdef ET_forest *ET_forest_build(ET_problem *problem, ET_params *parmas)

    ctypedef struct ET_params:
        
        uint32_t number_of_features_tested
        uint32_t number_of_trees
        bool regression
        uint32_t min_split_size
        bool select_features_with_replacement

    ctypedef struct class_probability:
        double label
        double probability

    ctypedef struct class_probability_vec:
        size_t n, m
        class_probability *a

    cdef double ET_forest_predict(ET_forest *forest, float *vector)
    cdef double ET_forest_predict_regression(ET_forest *forest,
                                             float *vector,
                                             uint32_t curtail_min_size)
    cdef double ET_forest_predict_class_majority(ET_forest *forest,
                                             float *vector,
                                             uint32_t curtail_min_size)
    cdef double ET_forest_predict_class_bayes(ET_forest *forest,
                                             float *vector,
                                             uint32_t curtail_min_size,
                                             bool smooth)
    cdef class_probability_vec *ET_forest_predict_probability(ET_forest *forest,
                                             float *vector,
                                             uint32_t curtail_min_size,
                                             bool smooth)
    cdef double *ET_forest_neighbors(ET_forest *forest,
                                             float *vector,
                                             uint32_t curtail_min_size)
    cdef double *ET_forest_feature_importance(ET_forest *forest,
                                             uint32_t curtail_min_size)

    ctypedef struct uchar_vec:
        size_t n, m
        unsigned char *a

    cdef void ET_tree_dump(ET_tree tree, uchar_vec *buffer)
    cdef void ET_forest_dump(ET_forest *forest, uchar_vec *buffer,
                                                bool with_trees)
    
    cdef ET_tree ET_tree_load(unsigned char **bufferp)
    cdef ET_forest *ET_forest_load(unsigned char **bufferp)
