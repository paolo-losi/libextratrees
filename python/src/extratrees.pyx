cimport cython
cimport numpy as np
from libc.stdlib cimport malloc, free
from cextratrees cimport ET_problem, ET_problem_destroy


cdef class Problem:

    cdef ET_problem *_prob

    def __dealloc__(self):
        if self._prob:
            ET_problem_destroy(self._prob)
            free(self._prob)


cdef Problem problem_factory(ET_problem *prob):
    cdef Problem instance = Problem.__new__(Problem)
    instance._prob = prob
    return instance


@cython.boundscheck(False)
@cython.wraparound(False)
def convert_to_problem(np.ndarray[np.float64_t, ndim=2] X,
                       np.ndarray[np.float64_t, ndim=1] y):

    if y.shape[0] != X.shape[0]:
        raise ValueError('y.shape[0] != X.shape[0]')

    cdef size_t n_features, n_samples
    cdef int i, j

    cdef ET_problem *cprob = <ET_problem *> malloc(sizeof(ET_problem))
    
    n_features = X.shape[1]
    n_samples  = X.shape[0]

    cprob.n_features = n_features
    cprob.n_samples = n_samples

    cprob.vectors = <float *> malloc(sizeof(float) * n_features * n_samples)
    cprob.labels = <double *> malloc(sizeof(double) * n_samples)

    for j in xrange(n_features):
        for i in xrange(n_samples):
            cprob.vectors[j * n_samples + i] = X[i, j]

    for i in xrange(n_samples):
        cprob.labels[i] = y[i]

    return problem_factory(cprob)
