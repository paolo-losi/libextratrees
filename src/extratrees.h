# ifndef EXTRA_TREES_H
# define EXTRA_TREES_H

# include <stdint.h>
# include <stdio.h>
# include <stdbool.h>
# include <math.h>
# include "kvec.h"

typedef kvec_t(uint32_t) uint_vec;
typedef kvec_t(double) double_vec;
typedef kvec_t(unsigned char) uchar_vec;

// --- problem ---

typedef struct {
    float *vectors;
    double *labels;
    uint32_t n_features;
    uint32_t n_samples;
} ET_problem;


// --- params ---

typedef struct ET_params {
    uint32_t number_of_features_tested;
    uint32_t number_of_trees;
    bool regression;
    uint32_t min_split_size;
    bool select_features_with_replacement;
} ET_params;


# define EXTRA_TREE_DEFAULT_CLASSIF_PARAMS(prob, params) do {            \
    (params).number_of_features_tested = ceil(sqrt((prob).n_features));  \
    (params).number_of_trees           = 100;                            \
    (params).regression                = false;                          \
    (params).min_split_size            = 1;                              \
    (params).select_features_with_replacement = false;                   \
    } while(0)

# define EXTRA_TREE_DEFAULT_REGR_PARAMS(prob, params) do {               \
    (params).number_of_features_tested = (prob).n_features;              \
    (params).number_of_trees           = 100;                            \
    (params).regression                = true;                           \
    (params).min_split_size            = 1;                              \
    (params).select_features_with_replacement = false;                   \
    } while(0)


// --- tree ---

#define ET_LEAF_NODE 'L'
#define ET_SPLIT_NODE 'S'

typedef struct ET_base_node {
    char type;
    uint32_t n_samples;
    double diversity;
} ET_base_node;

typedef struct ET_split_node {
    ET_base_node base;
    float threshold;
    uint32_t feature_id;
    ET_base_node *lower_node, *higher_node;
} ET_split_node;

typedef struct ET_leaf_node {
    ET_base_node base;
    uint_vec indexes;
} ET_leaf_node;

typedef ET_base_node *ET_tree;

// --- forest ---

#ifndef ET_CLASS_COUNTER
#define ET_CLASS_COUNTER
typedef struct ET_class_counter_struct ET_class_counter;
#endif

typedef struct {
    uint32_t n_features;
    uint32_t n_samples;
    kvec_t(ET_tree) trees;
    ET_params params;
    double *labels;
    ET_class_counter *class_frequency;
} ET_forest;


// --- predict types ---

typedef struct {
    uint32_t key;
    double weight;
} neighbour_weight;

typedef kvec_t(neighbour_weight) neighbour_weight_vec;


typedef struct {
    double label;
    double probability;
} class_probability;

typedef kvec_t(class_probability) class_probability_vec;


// --- functions ---

ET_problem *ET_load_libsvm_file(char *fname);
ET_forest *ET_forest_build(ET_problem *prob, ET_params *params);
double_vec *ET_forest_feature_importance(ET_forest *forest);
void ET_forest_destroy(ET_forest *forest);
void ET_forest_dump(ET_forest *forest, uchar_vec *buffer);
void ET_problem_print(ET_problem *prob, FILE *f);
void ET_problem_destroy(ET_problem *prob);

double ET_forest_predict(ET_forest *forest, float *vector);
double ET_forest_predict_regression(ET_forest *forest, float *v,
                                    uint32_t curtail_min_size);
double ET_forest_predict_class_majority(ET_forest *forest, float *v,
                                        uint32_t curtail_min_size);
double ET_forest_predict_class_bayes(ET_forest *forest, float *v,
                                     uint32_t curtail_min_size, bool smooth);
neighbour_weight_vec *ET_forest_neighbors(ET_forest *forest, float *vector,
                                          uint32_t curtail_min_size);
class_probability_vec *ET_forest_predict_probability(ET_forest *forest,
                                                    float *vector,
                                                    uint32_t curtail_min_size,
                                                    bool smooth);

# endif
