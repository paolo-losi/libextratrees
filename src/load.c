#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>

#include "rt.h"
#include "load.h"
#include "log.h"
#include "util.h"


char *new_err_str(char *msg, int line_count, char *line) {
    char *error_str;

    line[20] = '\0';
    error_str = malloc(1024);
    snprintf(error_str, 1024,
             "parsing error on line %d: %s. line=%s...",
             line_count, msg, line);
    return error_str;
}


int parse_line(parse_state *ps, char *line, int line_count) {
    int feat_id, n;
    double val;

    if (sscanf(line, "%lf%n", &val, &n) != 1) {
        ps->on_error(ps, new_err_str("label not found", line_count, line));
        return -1;
    }

    line += n;
    ps->on_new_label(ps, val);

    while(sscanf(line, "%d : %lf%n", &feat_id, &val, &n) == 2) {
        line += n;
        ps->on_new_feature(ps, feat_id, val);
    }

    while(isspace(*line)) line++;

    if (*line != '\0') {
        ps->on_error(ps, new_err_str("feature not found", line_count, line));
        return -1;
    }

    return 0;
}


void parse_libsvm_file(parse_state *ps, FILE *f, int bufsize) {
    char *buffer, *bp;
    int line_count = 0;

    rewind(f);
    bp = buffer = malloc(bufsize);
    check_mem(buffer);

    while(1) {
        int c = fgetc(f);
        if (c == '\n' || c == EOF) {
            if (bp != buffer) {
                line_count += 1;
                *bp = '\0';
                if (parse_line(ps, buffer, line_count)) break;
                bp = buffer;
            }
            if (c == EOF) break;
        } else {
            if ((bp - buffer) >= bufsize) {
                size_t newbufsize = bufsize * 4;
                buffer = realloc(buffer, newbufsize);
                check_mem(buffer);
                bp = buffer + bufsize;
                bufsize = newbufsize;
            }
            *bp++ = c;
        }
    }
    free(buffer);
    return;

    exit:
    if (buffer) free(buffer);
    ps->on_error(ps, "malloc failed");
    return;
}


// ----- size_parser -----

typedef struct size_parser {
    parse_state ops;
    char *error_msg;
    int n_samples, n_features;
} size_parser;

void size_parser_on_error(parse_state *ps, char *error_msg) {
    size_parser *sp = (size_parser *) ps;
    sp->error_msg = error_msg;
}

void size_parser_on_new_label(parse_state *ps, double val) {
    UNUSED(val);
    size_parser *sp = (size_parser *) ps;
    sp->n_samples += 1;
}

void size_parser_on_new_feature(parse_state *ps, int fid, double val) {
    UNUSED(val);
    size_parser *sp = (size_parser *) ps;
    if (fid > sp->n_features) 
        sp->n_features = fid;
}

void size_parser_init(size_parser *sp) {
    sp->ops.on_error = size_parser_on_error;
    sp->ops.on_new_label = size_parser_on_new_label;
    sp->ops.on_new_feature = size_parser_on_new_feature;
    sp->n_features = 0;
    sp->n_samples = 0;
    sp->error_msg = NULL;
}

void size_parser_destroy(size_parser *sp) {
    if(sp->error_msg) free(sp->error_msg);
}


// ----- load_parser -----

typedef struct load_parser {
    parse_state ops;
    int error_flag;
    float *vectors;
    double *labels;
    int sample_idx;
    int n_samples;
    int n_features;
} load_parser;

void load_parser_on_error(parse_state *ps, char *error_msg) {
    UNUSED(error_msg);
    load_parser *lp = (load_parser *) ps;
    lp->error_flag = 1;
}

void load_parser_on_new_label(parse_state *ps, double val) {
    load_parser *lp = (load_parser *) ps;
    lp->sample_idx += 1;
    lp->labels[lp->sample_idx] = val;
}

void load_parser_on_new_feature(parse_state *ps, int fid, double val) {
    load_parser *lp = (load_parser *) ps;
    lp->vectors[lp->sample_idx + lp->n_samples * (fid-1)] = val;
}

int load_parser_init(load_parser *lp, int n_features, int n_samples) {
    float *vectors;
    double *labels;

    lp->ops.on_error = load_parser_on_error;
    lp->ops.on_new_label = load_parser_on_new_label;
    lp->ops.on_new_feature = load_parser_on_new_feature;
    lp->error_flag = 0;
    lp->sample_idx = -1;
    lp->n_samples = n_samples;
    lp->n_features = n_features;
    
    vectors = calloc(n_samples * n_features, sizeof(float));
    check_mem(vectors);
    labels = calloc(n_samples, sizeof(double));
    check_mem(labels);

    lp->vectors = vectors;
    lp->labels = labels;
    return 0;

    exit:
    lp->error_flag = 1;
    return -1;
}

void load_parser_destroy(load_parser* lp) {
    if(lp->error_flag) {
        if(lp->vectors) free(lp->vectors);
        if(lp->labels)  free(lp->labels);
    }
}


rt_problem *rt_load_libsvm_file(char *fname) {
    FILE *f;
    size_parser *sp = NULL;
    load_parser *lp = NULL;
    rt_problem *prob = NULL;

    f = fopen(fname, "r");
    check(f, "Could not open %s.", fname);

    sp = calloc(1, sizeof(size_parser));
    check_mem(sp);
    size_parser_init(sp);
    parse_libsvm_file((parse_state *) sp, f, PARSE_BUF_SIZE);
    check(!sp->error_msg, "%s", sp->error_msg);

    lp = calloc(1, sizeof(load_parser));
    check_mem(lp);
    check(! load_parser_init(lp, sp->n_features, sp->n_samples),
          "load_parser could not allocate enough memory.");

    parse_libsvm_file((parse_state *) lp, f, PARSE_BUF_SIZE);
    check(! lp->error_flag, "Unexpected failure in loading phase.");

    prob = malloc(sizeof(rt_problem));
    check_mem(prob);

    prob->vectors = lp->vectors;
    prob->labels  = lp->labels;
    prob->n_features = sp->n_features;
    prob->n_samples  = sp->n_samples;

    exit:
    if (sp) {
        size_parser_destroy(sp);
        free(sp);
    }
    if (lp) {
        load_parser_destroy(lp);
        free(lp);
    }
    if(f) fclose(f);
    return prob;
}
