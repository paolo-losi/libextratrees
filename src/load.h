#ifndef RT_LOAD_H
#define RT_LOAD_H

#define PARSE_BUF_SIZE 4096 * 1024


typedef struct parse_state {
    void (*on_new_label) (struct parse_state *state,  double);
    void (*on_new_feature) (struct parse_state *state, int, double);
    void (*on_error) (struct parse_state *state, char *);
} parse_state;


void parse_libsvm_file(parse_state *ps, FILE *f, int bufsize); 

#endif
