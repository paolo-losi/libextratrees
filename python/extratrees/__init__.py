from .cbindings import convert_to_problem, convert_from_problem, train, load
