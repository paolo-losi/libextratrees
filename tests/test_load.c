#include "stdio.h"

#include "extratrees.h"
#include "load.h"
#include "test.h"

void load_simple_file() {
    test_header();
    ET_problem *prob;
    prob = ET_load_libsvm_file("test1.libsvm");
    ET_problem_print(prob, stderr);
}

void load_unexistent_file() {
    test_header();
    ET_problem *prob;
    prob = ET_load_libsvm_file("foo");
}

int main() {
    load_simple_file();
    load_unexistent_file();
    return 0;
}
