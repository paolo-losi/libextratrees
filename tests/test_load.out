>>> test: load_simple_file
problem: samples=2 features=3
sample 0. label=1 features=10 20 0
sample 1. label=2.3 features=0 0.1 1
>>> test: load_unexistent_file
[ERROR]      src/load.c:196  errno=No such file or directory. Could not open foo.
